protocols/netlink/netlink_reader.go

```diff
@@ -0,0 +1,226 @@
+	}
+}
+
+// create a path from a route
+func createPathFromRoute(r *netlink.Route) (*route.Path, error) {
+	nlPath, err := route.NewNlPathFromRoute(r, true)
+
+	if err != nil {
+		return nil, fmt.Errorf("Error while creating path object from route object", err)
+	}
+
+	return &route.Path{
+		Type:        route.NetlinkPathType,
+		NetlinkPath: nlPath,
+	}, nil
+}
+
+// propagate changes to all subscribed clients
+func (nr *NetlinkReader) propagateChanges(routes []netlink.Route) {
+	nr.removePathsFromClients(routes)
+	nr.addPathsToClients(routes)
+}
+
+// Add given paths to clients
+func (nr *NetlinkReader) addPathsToClients(routes []netlink.Route) {
+	for _, client := range nr.ClientManager.Clients() {
+		// only advertise changed routes
+
+		nr.mu.RLock()
+		advertise := route.NetlinkRouteDiff(routes, nr.routes)
+		nr.mu.RUnlock()
+
+		for _, r := range advertise {
+			// Is it a BIO-Written route? if so, skip it, dont advertise it
+			if r.Protocol == route.ProtoBio {
+				log.WithFields(routeLogFields(r)).Debug("Skipping bio route")
+				continue
+			}
+
+			// create pfx and path from route
+			pfx := bnet.NewPfxFromIPNet(r.Dst)
+			path, err := createPathFromRoute(&r)
+			if err != nil {
+				log.WithError(err).Error("Unable to create path")
+				continue
+			}
+
+			// Apply filter (if existing)
+			if nr.filter != nil {
+				var reject bool
+				// TODO: Implement filter that cann handle netlinkRoute objects
+				path, reject = nr.filter.ProcessTerms(pfx, path)
+				if reject {
+					log.Debug("Skipping route due to filter")
+					continue
+				}
+			}
+
+			log.WithFields(log.Fields{
+				"pfx":  pfx,
+				"path": path,
+			}).Debug("NetlinkReader - client.AddPath")
+			client.AddPath(pfx, path)
+		}
+	}
+}
+
+// Remove given paths from clients
+func (nr *NetlinkReader) removePathsFromClients(routes []netlink.Route) {
+	for _, client := range nr.ClientManager.Clients() {
+		// If there where no routes yet, just skip this funktion. There's nothing to delete
+		nr.mu.RLock()
+		if len(nr.routes) == 0 {
+			nr.mu.RUnlock()
+			break
+		}
+
+		// only withdraw changed routes
+		withdraw := route.NetlinkRouteDiff(nr.routes, routes)
+		nr.mu.RUnlock()
+
+		for _, r := range withdraw {
+			// Is it a BIO-Written route? if so, skip it, dont advertise it
+			if r.Protocol == route.ProtoBio {
+				continue
+			}
+
+			// create pfx and path from route
+			pfx := bnet.NewPfxFromIPNet(r.Dst)
+			path, err := createPathFromRoute(&r)
+			if err != nil {
+				log.WithError(err).Error("Unable to create path")
+				continue
+			}
+
+			// Apply filter (if existing)
+			if nr.filter != nil {
+				var reject bool
+				// TODO: Implement filter that cann handle netlinkRoute objects
+				path, reject = nr.filter.ProcessTerms(pfx, path)
+				if reject {
+					continue
+				}
+			}
+
+			log.WithFields(log.Fields{
+				"pfx":  pfx,
+				"path": path,
+			}).Debug("NetlinkReader - client.RemovePath")
+			client.RemovePath(pfx, path)
```

Comment:
yea. looks much simpler now and could reduce the complexity from `O(n*m)` to `O(n)` in the best case, where `n` stands for the number of routes and `m` for the number of clients, because i moved the `for client...` down to the place where the client is really needed.

Reply:
would like to click "Resolve conversation" if you think it's better now.
Please have a look at the refactored code :)